modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py

```diff
@@ -93,76 +119,47 @@ def apply(self, func, *args, **kwargs):
         -----
         It does not matter if `func` is callable or an ``ray.ObjectRef``. Ray will
         handle it correctly either way. The keyword arguments are sent as a dictionary.
+
+        If ``LazyExecution`` is enabled, the function is not applied immediately,
+        but is added to the execution tree.
         """
+        de = DeferredExecution(self._data_ref, func, args, kwargs)
+        if LazyExecution.get():
+            return self.__constructor__(de)
         log = get_logger()
         self._is_debug(log) and log.debug(f"ENTER::Partition.apply::{self._identity}")
-        data = self._data
-        call_queue = self.call_queue + [[func, args, kwargs]]
-        if len(call_queue) > 1:
-            self._is_debug(log) and log.debug(
-                f"SUBMIT::_apply_list_of_funcs::{self._identity}"
-            )
-            result, length, width, ip = _apply_list_of_funcs.remote(
-                data, *deconstruct_call_queue(call_queue)
-            )
-        else:
-            # We handle `len(call_queue) == 1` in a different way because
-            # this dramatically improves performance.
-            func, f_args, f_kwargs = call_queue[0]
-            result, length, width, ip = _apply_func.remote(
-                data, func, *f_args, **f_kwargs
-            )
-            self._is_debug(log) and log.debug(f"SUBMIT::_apply_func::{self._identity}")
+        data, meta, meta_offset = de.exec()
         self._is_debug(log) and log.debug(f"EXIT::Partition.apply::{self._identity}")
-        return self.__constructor__(result, length, width, ip)
+        return self.__constructor__(data, meta=meta, meta_offset=meta_offset)
+
+    @_inherit_docstrings(PandasDataframePartition.add_to_apply_calls)
+    def add_to_apply_calls(self, func, *args, length=None, width=None, **kwargs):
+        return self.__constructor__(
+            data=DeferredExecution(self._data_ref, func, args, kwargs),
+            length=length,
+            width=width,
+        )
 
+    @_inherit_docstrings(PandasDataframePartition.drain_call_queue)
     def drain_call_queue(self):
-        """Execute all operations stored in the call queue on the object wrapped by this partition."""
+        data = self._data_ref
+        if not isinstance(data, DeferredExecution):
+            return data
```

Comment:
What if we called multiple operations with add_to_apply_calls and then wanted to drain the call queue? I would expect all lazy operations to be performed but we just return the initial obj_ref.

Reply:
If the initial obj_ref was a DeferredExecution, in drain_call_queue() it's replaced with object ref. drain_call_queue()  triggers execution of the entire chain.

Reply:
I see, thanks!